File: BookStore_iOS/BookStore_CarlosNeto/BookStore_CarlosNeto/Controller/DetailViewController.m
//
//  DetailViewController.m
//  BookStore_CarlosNeto
//
//  Created by Carlos Neto on 16/04/2019.
//  Copyright © 2019 Carlos Neto. All rights reserved.
//

#import "DetailViewController.h"

@interface DetailViewController ()

@property (weak, nonatomic) IBOutlet UILabel *titleLbl;
@property (weak, nonatomic) IBOutlet UILabel *authorLbl;
@property (weak, nonatomic) IBOutlet UILabel *infoLbl;
@property (weak, nonatomic) IBOutlet UIButton *linkBtn;

@property (weak, nonatomic) IBOutlet UIView *titleView;
@property (weak, nonatomic) IBOutlet UIView *authorView;
@property (weak, nonatomic) IBOutlet UIView *infoView;
@property (weak, nonatomic) IBOutlet UIView *linkView;

@property CNItem *item;

@end

@implementation DetailViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    self.title = @"Book Item";
    
    if (!_item) {
        UINavigationController *nav = [[(UINavigationController *)self.splitViewController viewControllers] lastObject];
        [nav popViewControllerAnimated:YES];
    }
    _titleLbl.text = @"";
    _authorLbl.text = @"";
    _infoLbl.text = @"";
    [_linkBtn setTitle:@"" forState:UIControlStateNormal];
    [_titleView setHidden:YES];
    [_authorView setHidden:YES];
    [_infoView setHidden:YES];
    [_linkView setHidden:YES];
    _linkBtn.contentHorizontalAlignment = UIControlContentHorizontalAlignmentLeft;
}

- (void)setupUI:(CNItem*)cnItem
{
    _item = cnItem;
    
    _titleLbl.text = _item.volumeInfo.title;
    _authorLbl.text = _item.volumeInfo.authors.firstObject;
    _infoLbl.text = _item.volumeInfo.theDescription;
    [_linkBtn setTitle:_item.volumeInfo.infoLink forState:UIControlStateNormal];
    
    [_titleView setHidden:!_titleLbl.text.length];
    [_authorView setHidden:!_authorLbl.text.length];
    [_infoView setHidden:!_infoLbl.text.length];
    [_linkView setHidden:!_linkBtn.titleLabel.text.length];
}

- (IBAction)linkBtnAction:(UIButton *)sender
{
    NSURL *url = [NSURL URLWithString:_linkBtn.titleLabel.text];
    [[UIApplication sharedApplication] openURL:url];
}

@end

File: BookStore_iOS/BookStore_CarlosNeto/BookStore_CarlosNeto/Controller/DetailViewController.swift
//
//  DetailViewController.swift
//  BookStore_CarlosNeto
//
//  Created by Carlos Neto on 02/03/2020.
//  Copyright © 2020 Carlos Neto. All rights reserved.
//

import UIKit

class DetailViewController: UIViewController {
    
    @IBOutlet private weak var favBtn: UIButton!
    @IBOutlet private weak var titleLbl: UILabel!
    @IBOutlet private weak var authorLbl: UILabel!
    @IBOutlet private weak var infoLbl: UILabel!
    @IBOutlet private weak var linkBtn: UIButton!
    @IBOutlet private weak var favView: UIView!
    @IBOutlet private weak var titleTag: UILabel!
    @IBOutlet private weak var authorTag: UILabel!
    @IBOutlet private weak var infoTag: UILabel!
    @IBOutlet private weak var linkView: UIView!
    
    private var item: Item?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        title = "Book Item"
        if item == nil, let nav = splitViewController?.viewControllers.last as? UINavigationController {
            nav.popViewController(animated: true)
        }
        titleLbl.text = ""
        authorLbl.text = ""
        infoLbl.text = ""
        linkBtn.setTitle("", for: .normal)
        linkBtn.contentHorizontalAlignment = .right
        setTags()
    }
    
    func setupUI(item: Item?) {
        self.item = item
        setFavorite(isFavorite: self.item?.isFavorite ?? false)
        let addToFavorite = UIImage(named: "addToFavorite")
        let addedFavorite = UIImage(named: "addedFavorite")
        favBtn.setImage(addToFavorite, for: .normal)
        favBtn.setImage(addedFavorite, for: .selected)
        if let identifier = self.item?.identifier {
            favBtn.accessibilityIdentifier = "\(identifier)_detail"
        } else {
            favBtn.accessibilityIdentifier = nil
        }
        titleLbl.text = item?.volumeInfo?.title
        authorLbl.text = item?.volumeInfo?.authors?.first
        infoLbl.text = item?.volumeInfo?.description
        linkBtn.setTitle(item?.volumeInfo?.infoLink, for: .normal)
        
        setTags()
    }
    
    private func setTags() {
        titleTag.isHidden = titleLbl.text?.isEmpty ?? true
        authorTag.isHidden = authorLbl.text?.isEmpty ?? true
        infoTag.isHidden = infoLbl.text?.isEmpty ?? true
    }

    @IBAction private func favBtnAction(_ sender: UIButton) {
        setFavorite(isFavorite: !(item?.isFavorite ?? true))
    }

    @IBAction private func linkBtnAction(_ sender: UIButton) {
        guard let url = URL(string: linkBtn.titleLabel?.text ?? ""), !url.absoluteString.isEmpty else { return }
        UIApplication.shared.open(url, options: [:], completionHandler: { success in
            print("Open \(url.absoluteString): \(success)")
        })
    }
    
    private func setFavorite(isFavorite: Bool) {
        item?.isFavorite = isFavorite
        favBtn.isSelected = isFavorite
        if let identifier = item?.identifier,
            let btn = UIButton.firstWith(identifier: "\(identifier)_master") {
            btn.isSelected = isFavorite
        }
    }
}
